### ISSUE TITLE
poetry run raises FileNotFoundError without any more information (workaround found)

### ISSUE THREAD

**Issue by ThmX** [python-poetry/poetry#3592]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: docker alpine:3.13
- **Poetry version**: 1.1.4
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: vanilla project from `poetry init` with a main.py and the following `scripts` section:
```toml
[tool.poetry.scripts]
main = "main:main"
```

## Issue

Using the following Dockerfile as base image on our build pipeline.

```docker
FROM alpine:3.13
RUN apk add --update --no-cache python3 python3-dev py3-pip py3-pandas py3-numpy-dev hdf5-dev
RUN pip install poetry
RUN poetry config virtualenvs.create false
```

`poetry run main` crashes with the following error:

```
/workdir # poetry init
/workdir # poetry run -vvv main

  FileNotFoundError

  [Errno 2] No such file or directory

  at /usr/lib/python3.8/os.py:601 in _execvpe
       597│         path_list = map(fsencode, path_list)
       598│     for dir in path_list:
       599│         fullname = path.join(dir, file)
       600│         try:
    →  601│             exec_func(fullname, *argrest)
       602│         except (FileNotFoundError, NotADirectoryError) as e:
       603│             last_exc = e
       604│         except OSError as e:
       605│             last_exc = e
```

> Note that enabling or disabling `virtualenvs.create` has the same effect.

Long story short, I've been able to work around the issue and hope this might help others. Here is what did the trick:
```bash
/workdir # ln -s /usr/bin/python3 /usr/bin/python
```

I suppose that `poetry run` is expecting `python` to be found in the PATH on contrary to `poetry` which is directly using `python3`:
```bash
/workdir # head -1 /usr/bin/poetry
#!/usr/bin/python3
```

**Comment by sziem** [comment]
I have encountered the same error, but found another workaround.  I put the script i wanted to run in a subfolder (such as src).  For example, `poetry run script.py` did not work, but `poetry run src/script.py` did.

I have also noticed that the same error occurs on my system, if I just run 
`poetry run --help`, which according to `poetry help run` is supposed to work.

**Comment by dephekt** [comment]
I started having this issue today (seemingly randomly, Poetry was working fine a few days ago).

None of the above workarounds have helped. It's breaking every poetry project I have, not just a specific one.

If I `source ~.cache/pypoetry/virtualenvs/virustotal-4DyFcNxD-py3.8/bin/activate` and then run `python run.py --help` my script works fine.

I added a print in the os package on L600 just before the try/except clause for the call to exec_func() and see that it's looking in poetry's bin directory for my script, which obviously is not going to be there:

```
$ poetry run run.py --help
b'/home/dephekt/.cache/pypoetry/virtualenvs/virustotal-4DyFcNxD-py3.8/bin/run.py' ['run.py', '--help']
b'/home/dephekt/bin/run.py' ['run.py', '--help']
b'/usr/local/bin/run.py' ['run.py', '--help']
b'/home/dephekt/.local/bin/run.py' ['run.py', '--help']
b'/home/dephekt/.poetry/bin/run.py' ['run.py', '--help']
b'/home/dephekt/.local/bin/run.py' ['run.py', '--help']
b'/home/dephekt/bin/run.py' ['run.py', '--help']
b'/usr/local/sbin/run.py' ['run.py', '--help']
b'/usr/local/bin/run.py' ['run.py', '--help']
b'/usr/sbin/run.py' ['run.py', '--help']
b'/usr/bin/run.py' ['run.py', '--help']
b'/sbin/run.py' ['run.py', '--help']
b'/bin/run.py' ['run.py', '--help']
b'/usr/games/run.py' ['run.py', '--help']
b'/usr/local/games/run.py' ['run.py', '--help']
b'/snap/bin/run.py' ['run.py', '--help']

  FileNotFoundError

  [Errno 2] No such file or directory

  at /usr/lib/python3.8/os.py:602 in _execvpe
       598│     for dir in path_list:
       599│         fullname = path.join(dir, file)
       600│         print(fullname, *argrest)
       601│         try:
    →  602│             exec_func(fullname, *argrest)
       603│         except (FileNotFoundError, NotADirectoryError) as e:
       604│             last_exc = e
       605│         except OSError as e:
       606│             last_exc = e
```

**Comment by iotanbo** [comment]
Tried `poetry` on Ubuntu20.04, same problem. After installing it `pyenv` stopped working as intended. 
Before:
```bash
python --version 
# Python 3.8.5
python3.9 --version 
# Python 3.9.5
```
After:
```bash
python --version
# Python 3.8.5
python3 --version
# Python 3.8.5
python3.9 --version
# Command 'python3.9' not found, did you mean: ...
which python
# /usr/bin/python
```

**Comment by iotanbo** [comment]
I fixed the `pyenv` issue, but running `poetry run -h` yields following result:

```bash
$ poetry run -h

  FileNotFoundError

  [Errno 2] No such file or directory: b'/snap/bin/-h'

  at ~/.pyenv/versions/3.8.6/lib/python3.8/os.py:601 in _execvpe
       597│         path_list = map(fsencode, path_list)
       598│     for dir in path_list:
       599│         fullname = path.join(dir, file)
       600│         try:
    →  601│             exec_func(fullname, *argrest)
       602│         except (FileNotFoundError, NotADirectoryError) as e:
       603│             last_exc = e
       604│         except OSError as e:
       605│             last_exc = e
```
My PATH variable looks as follows:
```bash
/home/name/.pyenv/plugins/pyenv-virtualenv/shims:/home/name/.local/bin:/home/name/.pyenv/plugins/pyenv-virtualenv/shims:/home/name/.pyenv/shims:/home/name/.pyenv/bin:/home/name/.cargo/bin:/home/name/.local/bin:/home/name/.pyenv/shims:/home/name/.pyenv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```

**Comment by iotanbo** [comment]
Finally I fixed my problem: it was the bad entry in the `pyproject.toml`:

```toml
[tool.poetry.scripts]
poetry-test = "poetry_test.console:main"
```
Here `poetry-test` is the script to be executed like
`poetry run poetry-test`, `poetry_test` is the package name,
`console.py` is a python file inside the package, and `main` is the function inside `console.py`.
 Sorry for the newbie's comments, I've just got acquainted with `poetry`  :). Hopefully this will help someone else.

**Comment by epogrebnyak** [comment]
Same issue here of Ubuntu:

```
root@LAPTOP-E0HT1CM9:/mnt/c/Users/epogr/Documents/GitHub/rides-blueprint# poetry run --help

  FileNotFoundError

  [Errno 2] No such file or directory: b'/snap/bin/--help'

  at /usr/lib/python3.8/os.py:601 in _execvpe
       597│         path_list = map(fsencode, path_list)
       598│     for dir in path_list:
       599│         fullname = path.join(dir, file)
       600│         try:
    →  601│             exec_func(fullname, *argrest)
       602│         except (FileNotFoundError, NotADirectoryError) as e:
       603│             last_exc = e
       604│         except OSError as e:
       605│             last_exc = 
```

**Comment by rustymagnet3000** [comment]
> poetry run -vvv main

should that be?

`poetry run python -vvv main`

**Comment by DishaDudhal** [comment]
Facing dvc issue with following: Note that I have already exported the path variable in my .bashrc file 

dd$dd:~ poetry run dvc init --subdir

  FileNotFoundError

  [Errno 2] No such file or directory: b'/home/dd/.local/bin/dvc'

  at /usr/lib/python3.8/os.py:601 in _execvpe
       597│         path_list = map(fsencode, path_list)
       598│     for dir in path_list:
       599│         fullname = path.join(dir, file)
       600│         try:
    →  601│             exec_func(fullname, *argrest)
       602│         except (FileNotFoundError, NotADirectoryError) as e:
       603│             last_exc = e
       604│         except OSError as e:
       605│             last_exc = e

**Comment by NK74** [comment]
Up ;)

**Comment by wagnerluis1982** [comment]
I just noticed in the description issue that is not following the CI recommendations to use poetry.

What ringed alarms to me is that you're installing poetry in the global python lib, not in some isolated environment (provided by `venv`, `pipx` or official installer).

I believe if you follow, you're not going to have the issue.

Link: https://python-poetry.org/docs/#ci-recommendations
